File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Solution {
    public int solution(int[,] maps) {
        int n = maps.GetLength(0);
        int m = maps.GetLength(1);
        
        int[,] visited = new int[n,m];
        Queue<(int x, int y)> q = new Queue<(int, int)>();
        
        int[] dx = {-1,1,0,0};
        int[] dy = {0,0,-1,1};
        
        q.Enqueue((0,0));
        visited[0,0] = 1;
        
        while (q.Count > 0)
        {
            var (x,y) = q.Dequeue();
            
            for(int i = 0; i < 4; i++)
            {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (0 <= nx && nx < n && 0 <= ny && ny < m && maps[nx,ny] == 1 && visited[nx,ny] == 0)
                {
                    q.Enqueue((nx,ny));
                    visited[nx,ny] = visited[x,y] + 1;
                }
            }
        }
        if(visited[n-1,m-1] == 0)
            return -1;
        else
            return visited[n-1,m-1];
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    
    q =deque()
    q.append((0,0))
    
    while q:
        x,y = q.popleft()
        
        for i in range(4):
            nx = x+ dx[i]
            ny = y+ dy[i]
            
            if 0 <= nx < n and 0 <= ny < m and maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                q.append((nx,ny))
             
    if maps[n-1][m-1] == 1:
        return -1
    else:
        return maps[n-1][m-1]
